Add SPI to decode image data with size

szewai opened the pull request:
#### 8e8ab054e9b335906ef4cc8a8bfb9f0940fccdb0
<pre>
Add SPI to decode image data with size
<a href="https://bugs.webkit.org/show_bug.cgi?id=281776">https://bugs.webkit.org/show_bug.cgi?id=281776</a>
<a href="https://rdar.apple.com/136706482">rdar://136706482</a>

Reviewed by Darin Adler.

Provide a way for client to generate icon data and decode image data in a sandboxed process.
New API Test: WebKit.CreateIconFromImageData.

* Source/WebCore/platform/graphics/ImageUtilities.h:
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::createNativeImageFromData):
(WebCore::cfString):
(WebCore::expandNativeImageToData):
(WebCore::createIconDataFromImageData):
(WebCore::decodeImageWithSize):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _createIconDataFromImageData:withLengths:completionHandler:]):
(-[WKWebView _decodeImageData:preferredSize:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::createIconDataFromImageData):
(WebKit::WebPageProxy::decodeImageData):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::createIconDataFromImageData):
(WebKit::WebPage::decodeImageData):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageData)):

Canonical link: <a href="https://commits.webkit.org/285530@main">https://commits.webkit.org/285530@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f06d3bea28a15959f9e8c46568475fa567d57585

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72899 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52324 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25699 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77097 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24133 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75014 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60129 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23949 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57324 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15806 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75966 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47303 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62745 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37745 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43964 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20213 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22471 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65805 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20569 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78770 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/17145 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19714 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65780 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/17194 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62751 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65047 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16077 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13356 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7013 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/48122 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2909 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/49189 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50484 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48934 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -125,4 +127,96 @@ Vector<String> transcodeImages(const Vector<String>& paths, const String& destin
+
+    RetainPtr image = imageDecoder->createFrameImageAtIndex(0);
+    if (!image)
+        return { };
```
I think this part of the function can be moved to a separate function called:
```
RefPtr<NativeImage> createNativeImageFromData(std::span<const uint8_t> data)
```

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -125,4 +127,96 @@ Vector<String> transcodeImages(const Vector<String>& paths, const String& destin
+    if (!CGImageDestinationFinalize(destination.get()))
+        return { };
+
+    return makeVector(destinationData.get());
```
And this part can be moved to a separate function called
```
Vector<uint8_t> expandNativeImageToData(NativeImage& image, const String& uti, std::span<const unsigned> lengths)
```

And your function can make calls to these functions after making sure the input data can be decoded and passing valid lengths (or sizes) for the icon image.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -125,4 +127,96 @@ Vector<String> transcodeImages(const Vector<String>& paths, const String& destin
+    }
+
+    if (!image)
+        image = imageDecoder->createFrameImageAtIndex(imageDecoder->primaryFrameIndex());
```
How about writing this code this way?
```
    auto preferredIndex = [&]() -> size_t {
        if (!preferredSize)
            return imageDecoder->primaryFrameIndex();
        for (size_t index = 0; index < frameCount; ++index) {
            if (imageDecoder->frameSizeAtIndex(index) == *preferredSize)
                return index;
        }
        return imageDecoder->primaryFrameIndex();
    };

    RetainPtr image = imageDecoder->createFrameImageAtIndex(preferredIndex());
```

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -125,4 +127,96 @@ Vector<String> transcodeImages(const Vector<String>& paths, const String& destin
+    return makeVector(destinationData.get());
+}
+
+RefPtr<WebCore::ShareableBitmap> decodeImageWithSize(std::span<const uint8_t> data, std::optional<WebCore::FloatSize> preferredSize)
```
Since this code is in WebCore, you do not need to include the namespace `WebCore`.

Also I think it is better to make this function return a `NativeImage` and leave transferring it to `ShareableBitmap` to the caller in WebKit.

Also including `decode` in the name makes it hard to know what this function is doing or returning. I would suggest naming this function something like
```
RefPtr<NativeImage> createNativeImageFromData(std::span<const uint8_t> data, std::optional<WebCore::FloatSize> preferredSize)
```

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +166,95 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+    RetainPtr colorSpace = adoptCF(CGColorSpaceCreateWithName(kCGColorSpaceGenericRGB));
+    RetainPtr destinationData = adoptCF(CFDataCreateMutable(0, 0));
+    String destinationUTI = "com.microsoft.ico"_s;
+    RetainPtr uti = destinationUTI.createCFString();
```
This could just be:
```
auto type = CFSTR("com.microsoft.ico");
```
Or maybe inline in the call to `CGImageDestinationCreateWithData`.

There’s no need to involve WTF in creating a `CFStringRef` for a string constant.

szewai replied:
Removed `String`.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +166,95 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+    for (auto length : targetLengths) {
+        RetainPtr context = adoptCF(CGBitmapContextCreate(nullptr, length, length, 8, length * 4, colorSpace.get(), kCGImageAlphaPremultipliedLast));
+        if (!context)
+            continue;
```
I don’t understand why this just skips a single length and keeps trying with the other lengths. Is that the right thing to do with the creation fails?

What about if all of these calls fail for all lengths? Do we have test cases covering these?

szewai replied:
Yes it's probably better to return early if a required length cannot be created. I don't know when this function can fail so not sure how to create test for it.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +166,95 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+        CGContextSetInterpolationQuality(context.get(), kCGInterpolationHigh);
+        CGContextDrawImage(context.get(), CGRectMake(0, 0, length, length), image.get());
+        if (RetainPtr newImage = adoptCF(CGBitmapContextCreateImage(context.get())))
+            CGImageDestinationAddImage(destination.get(), newImage.get(), nullptr);
```
Same question here. Is silently skipping one of the lengths the right thing to do?

szewai replied:
Will change to early return.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +166,95 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+
+    if (preferredSize) {
+        for (size_t index = 0; index < frameCount && !image; ++index) {
+            FloatSize frameSize = imageDecoder->frameSizeAtIndex(index);
```
Not sure we need a local variable for this.

szewai replied:
Removed.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +166,95 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+    if (!context)
+        return nullptr;
+
+    context->drawNativeImage(*nativeImage, FloatRect({ }, nativeImage->size()), FloatRect({ }, nativeImage->size()), { CompositeOperator::Copy });
```
Could make the `FloatRect` once in a local variable and use it twice.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +166,95 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+        return nullptr;
+
+    context->drawNativeImage(*nativeImage, FloatRect({ }, nativeImage->size()), FloatRect({ }, nativeImage->size()), { CompositeOperator::Copy });
+    return bitmap.releaseNonNull();
```
I think this should just be `return bitmap;`

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3289,6 +3289,45 @@ - (void)_getInformationFromImageData:(NSData *)imageData completionHandler:(void
+    targetLengths.reserveInitialCapacity(lengths.count);
+    for (NSNumber *length in lengths) {
+        if (length && length.unsignedIntValue)
+            targetLengths.append(length.unsignedIntValue);
```
Given how Objective-C works, we don’t need to check `length` for nil, and we also don’t want to call the method twice.
```
if (unsigned lengthValue = length.unsignedIntValue)
    targetLengths.append(lengthValue);
```

darinadler commented on Source/WebCore/platform/graphics/ImageUtilities.h:
```diff
@@ -51,6 +51,8 @@ enum class ImageDecodingError : uint8_t {
 };
 WEBCORE_EXPORT String descriptionString(ImageDecodingError);
 WEBCORE_EXPORT Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailableSizesFromImageData(std::span<const uint8_t>);
+WEBCORE_EXPORT Vector<uint8_t> createIconFromImageData(std::span<const uint8_t> data, std::span<const unsigned> lengths);
```
Unfortunate to move this from a `CFDataRef` into a `Vector` then ship it across the process boundary and then turn the second `Vector` into an `NSData`. Basically copying the data two extra times. This is platform-specific code, but we are round-tripping unnecessarily through a platform-independent data type. Would be cleaner to avoid that. Maybe our IPC knows how to deal with `CFData`/`NSData`?

szewai replied:
I think we might avoid this by using std::span in message; will try

cdumez replied:
I believe you can construct a `SharedBuffer` from your CFData/NSData and pass a the `SharedBuffer` over IPC. Not only will it use span for encoding in the worst case, it may also send shared memory if the CFData/NSData is mmap'd (which would be even more efficient).

szewai replied:
ok; will update to `SharedBuffer`

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +166,95 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+    if (frameCount != 1)
+        return { };
+
+    RetainPtr image = imageDecoder->createFrameImageAtIndex(0);
```
Should not we get the frame of the primary index instead of the first frame?

szewai replied:
I think primary index equals to first index since the image should only contains one frame; but I think using primary index and removing the requirement of "one frame" would also work!

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +166,95 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
     return std::make_pair(WTFMove(uti), WTFMove(sizes));
 }
 
+Vector<uint8_t> createIconFromImageData(std::span<const uint8_t> data, std::span<const unsigned> lengths)
```
Can we be explicit about this function is returning? Since we are returning `IconData` not `Icon` and since we do not know if the passed data is an ImageData or just `Data`, can't we use something like `createIconDataFromData()`?

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +166,95 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+
+    size_t frameCount = imageDecoder->frameCount();
+    if (!frameCount)
+        return nullptr;
```
This code is repeated above. So I would suggest having these set of functions:
```
RefPtr<NativeImage> createNativeImageFromData(std::span<const uint8_t> data,
std::optional<WebCore::FloatSize> preferredSize = std::nullopt);
Vector<uint8_t> expandNativeImageToData(NativeImage& image, const String& uti, std::span<const unsigned> lengths);
Vector<uint8_t> createIconDataFromData(std::span<const uint8_t> data, std::span<const unsigned> lengths);
```
The function  `createNativeImageFromData()` will be called form `createIconDataFromData()` and form `WebPage::decodeImageData()`.

szewai replied:
Thanks for the review and suggestions! I updated the functions as suggested.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +167,90 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+    if (!isSupportedImageType(sourceUTI))
+        return nullptr;
+
+    size_t frameCount = imageDecoder->frameCount();
```
Should do this right before it's used rather than outside the lambda.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +167,90 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+{
+    RetainPtr colorSpace = adoptCF(CGColorSpaceCreateWithName(kCGColorSpaceGenericRGB));
+    RetainPtr destinationData = adoptCF(CFDataCreateMutable(0, 0));
+    RetainPtr cfUTI = adoptCF(CFStringCreateWithCStringNoCopy(kCFAllocatorDefault, uti.characters(), kCFStringEncodingASCII, kCFAllocatorNull));
```
Seems like this idiom for turning an ASCIILiteral into a RetainPtr<CFStringRef> belongs in a helper function rather than being written in line like this. Don't need to move it into a header but I'd at least name it and make it separate here.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +167,90 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+    for (auto length : lengths) {
+        RetainPtr context = adoptCF(CGBitmapContextCreate(nullptr, length, length, 8, length * 4, colorSpace.get(), kCGImageAlphaPremultipliedLast));
+        if (!context)
+            return { };
```
Could use `nullptr`.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +167,90 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+        CGContextDrawImage(context.get(), CGRectMake(0, 0, length, length), image.platformImage().get());
+        RetainPtr newImage = adoptCF(CGBitmapContextCreateImage(context.get()));
+        if (!newImage)
+            return { };
```
Could use `nullptr`.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +167,90 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+    }
+
+    if (!CGImageDestinationFinalize(destination.get()))
+        return { };
```
Could use `nullptr`.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +167,90 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+    if (!nativeImage)
+        return { };
+
+    std::array<unsigned, 4> availableLengths { { 16, 32, 48, 256 } };
```
Should this have a comment about why these are the correct sizes?

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +167,90 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+
+static RefPtr<SharedBuffer> expandNativeImageToData(NativeImage& image, ASCIILiteral uti, std::span<const unsigned> lengths)
+{
+    RetainPtr colorSpace = adoptCF(CGColorSpaceCreateWithName(kCGColorSpaceGenericRGB));
```
Why is this the correct color space? Why not sRGB?

szewai replied:
sRGB should work too; I will change.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -164,4 +167,90 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
+        if (!context)
+            return { };
+
+        CGContextSetInterpolationQuality(context.get(), kCGInterpolationHigh);
```
This seems arbitrary and is not reflected in the function name. Maybe a comment about why it's correct?

szewai committed:
Add SPI to decode image data with size
https://bugs.webkit.org/show_bug.cgi?id=281776
rdar://136706482

Reviewed by Darin Adler.

Provide a way for client to generate icon data and decode image data in a sandboxed process.
New API Test: WebKit.CreateIconFromImageData.

* Source/WebCore/platform/graphics/ImageUtilities.h:
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::createNativeImageFromData):
(WebCore::cfString):
(WebCore::expandNativeImageToData):
(WebCore::createIconDataFromImageData):
(WebCore::decodeImageWithSize):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _createIconDataFromImageData:withLengths:completionHandler:]):
(-[WKWebView _decodeImageData:preferredSize:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::createIconDataFromImageData):
(WebKit::WebPageProxy::decodeImageData):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::createIconDataFromImageData):
(WebKit::WebPage::decodeImageData):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageData)):

Canonical link: https://commits.webkit.org/285530@main
